src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs

```diff
@@ -185,22 +214,36 @@ public void Add(string key, object value)
                 throw new ArgumentNullException(nameof(key));
             }
 
-            EnsureWritable();
-            InnerDictionary.Add(key, value);
+            Upgrade();
+
+            var list = ((ListStorage)_storage);
+            list.Inner.Add(new KeyValuePair<string, object>(key, value));
         }
 
         /// <inheritdoc />
         public void Clear()
         {
-            EnsureWritable();
-            InnerDictionary.Clear();
+            Upgrade();
+
+            var list = ((ListStorage)_storage);
+            list.Inner.Clear();
         }
 
         /// <inheritdoc />
         bool ICollection<KeyValuePair<string, object>>.Contains(KeyValuePair<string, object> item)
         {
-            EnsureWritable();
-            return ((ICollection<KeyValuePair<string, object>>)InnerDictionary).Contains(item);
+            Upgrade();
```

Comment:
Should exit early if storage is currently `EmptyStorage`. Particularly strange to upgrade for a read operation.